Add cets:insert_new operation

arcusfelis opened the pull request:
This operation uses one node in the cluster to verify that the inserted record is new.

# Reasoning

It is needed for mod_muc, because it tries to use room-name as a key.
There is a chance the room would be registered twice, so a transaction would be needed.
But insert-new looks like more simple solution (i.e. no need to invent locks for transaction).

# How it works

- we choose one node in the cluster where we would send an initial request.
- this node calls ets:insert_new and replicates the changes. 
- if ets:insert_new returns false - it does not replicate the changes and replies to the caller with `false` instead of `true`.

- We need to call `rpc:async_call(Node, ets, delete, [MonTab, Mon]);` to remove reference from the mon_tab... 


There is not much difference between insert and insert_new (2 nodes, 1 thread to test):

```
Name               ips        average  deviation         median         99th %
mnesia         21.90 K       45.65 μs    ±24.98%       42.80 μs      106.23 μs
ets            20.95 K       47.73 μs    ±76.92%       45.07 μs      100.62 μs
cets           19.65 K       50.89 μs    ±21.38%       50.28 μs       99.67 μs
cets_new       18.96 K       52.74 μs    ±29.13%       51.81 μs      101.72 μs
```

arcusfelis committed:
Fix calling CETS server on the remote node

arcusfelis committed:
Add cets:insert_new/2

arcusfelis committed:
Add failing path testcases

arcusfelis committed:
Add insert_new_when_inconsistent testcase

arcusfelis committed:
Optimize getting leader

Use ETS table instead of gen_server:call

arcusfelis committed:
Fix dialyzer

Reformat code

chrzaszcz commented on src/cets_call.erl:
```diff
@@ -83,8 +89,18 @@ wait_for_updated(Mon, {Servers, MonTab}) ->
+    ok.
+
+%% Edgecase: special treatment if Server is on the remote node
+delete_from_mon_tab({remote, Node, MonTab}, Mon) ->
```
It would be nice to have specs for the `{remote, Node, MonTab}` especially that this tuple is called just MonTab above, which is confusing.

chrzaszcz commented on src/cets_call.erl:
```diff
@@ -110,3 +126,24 @@ where(Name) when is_atom(Name) -> whereis(Name);
+    case Res of
+        {error, wrong_leader} ->
+            ?LOG_WARNING(#{what => wrong_leader, server => Server, operation => Op}),
+            timer:sleep(10),
```
From what I see, this might introduce super-long delays to the code using this function. It doesn't look good.

arcusfelis replied:
it is 10 milliseconds delay. But I could retry without delay, it will just make our local server less happy during the join operation (and extra load for the remote server too). Alternatively, I can wait for something like "join finished" event.

Will apply retry without delay, add a smarter solution in a separate PR.

chrzaszcz replied:
I misread as seconds, my bad. 
The only issue is that a bug could cause it to spin forever (especially with zero delay as in the latest commit), at least in theory. A retry counter (or a backoff) would suffice.

chrzaszcz commented on test/cets_SUITE.erl:
```diff
@@ -99,6 +108,141 @@ inserted_records_could_be_read_back_from_replicated_table(_Config) ->
+    ok = cets_join:join(join_lock1_insnew_back2, #{}, Pid1, Pid2),
+    Leader = cets:get_leader(Pid1),
+    cets:set_leader(Leader, false),
+    spawn(fun() ->
```
There is no pause, so `insert_new` will be _most likely_ executed before this code... this makes me wonder why we need this function at all.

arcusfelis replied:
oh, will add the comment.
Basically, in previous line we use `cets:set_leader(Leader, false)` which makes the call `cets:insert_new` never returning (it retries insert to the leader, but leader says "I am not a leader").

And this repeats until we actually execute code inside spawn function to fix the leader.

arcusfelis committed:
Add specs for wait_info

Add comments into insert_new_is_retried_when_leader_is_reelected
Do not wait before retry in send_leader_op

arcusfelis committed:
Add backoff in send_leader_op retries logic

Throw send_leader_op_failed error

chrzaszcz commented on src/cets_call.erl:
```diff
@@ -110,3 +128,36 @@ where(Name) when is_atom(Name) -> whereis(Name);
+        {error, wrong_leader} ->
+            ?LOG_WARNING(#{what => wrong_leader, server => Server, operation => Op}),
+            %% We are free to retry
+            %% While it is infinite retries, the leader election logic is simple.
```
It isn't infinite retries anymore

chrzaszcz commented on src/cets_call.erl:
```diff
@@ -110,3 +128,36 @@ where(Name) when is_atom(Name) -> whereis(Name);
+    case Res of
+        {error, wrong_leader} ->
+            ?LOG_WARNING(#{what => wrong_leader, server => Server, operation => Op}),
+            %% We are free to retry
```
For me this comment line is redundant.

chrzaszcz reviewed (approved):
Looks good :+1:

arcusfelis committed:
Edited a comment about retries in cets_call:send_leader_op
